openedx/features/course_experience/views/course_home_messages.py

```diff
@@ -107,36 +110,56 @@ def _register_course_home_messages(request, course, user_access, course_start_da
     """
     Register messages to be shown in the course home content page.
     """
-    if user_access['is_anonymous']:
-        CourseHomeMessages.register_info_message(
-            request,
-            Text(_(
-                '{sign_in_link} or {register_link} and then enroll in this course.'
-            )).format(
-                sign_in_link=HTML('<a href="/login?next={current_url}">{sign_in_label}</a>').format(
-                    sign_in_label=_('Sign in'),
-                    current_url=urlquote_plus(request.path),
+
+    is_course_public = allow_public_access(course, [COURSE_VISIBILITY_PUBLIC])
+    is_self_enrollment_allowed = course_open_for_self_enrollment(course.id)
+
+    if is_course_public and is_self_enrollment_allowed:
```

Comment:
We'll need to consider the case where the course maybe marked public but does not allow self enrollment. 
In that case, Enroll now button should not be shown. Instead, we can probably display a text like "You must be enrolled in the course to see full course content".
You can find a similar implementation here : https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/views/views.py#L531